### PR TITLE
Add Support for Domain Linkage

### PR DESCRIPTION
**Pull request by abdulmth**
# Description of change
- Add [DIDConfigurationResource](https://identity.foundation/.well-known/resources/did-configuration/#did-configuration-resource) and a convenient builder for [DomainLinkageCredential](https://identity.foundation/.well-known/resources/did-configuration/#domain-linkage-credential).
- Add domain linkage example.
- Domain Linkage Verifier.

## Links to any relevant issues
#154

## Type of change

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

### CODE REVIEW

File: identity_resolver/src/resolution/domain_linkage_verifier.rs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    let configuration_string: &str = include_str!("./tests/fixtures/did_configuration/config1.json");
+    let mut configuration: DIDConfigurationResource =
+      DIDConfigurationResource::from_json(configuration_string).unwrap();
+    let credential: &mut Credential = configuration.linked_dids_mut().get_mut(0).unwrap();
+    let (doc, _) = compose_did_document(credential.issuer.url().to_string());
+    let res = verify_config(doc, &configuration).await;
+    assert!(res.is_err());
+  }
+
+  // Todo: Add further tests for error cases.
```
Comment (PhilippGackstatter) [inline identity_resolver/src/resolution/domain_linkage_verifier.rs:248]: Indeed it would be great to verify that the cases that `verify_domain_linkage_credential` seems to catch are indeed caught. You can setup a general function that does the generic setup (load credential, create issuer document, create configuration resource) and then take a closure that modifies a small part of the test and verify the result with another closure.

File: identity_resolver/src/resolution/domain_linkage_verifier.rs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    let resolved_doc: AbstractThreadSafeValidatorDocument = self.resolver.resolve(issuer_did).await.map_err(|err| {
+      Error::new(ErrorCause::DNSVerificationError {
+        reason: format!("Issuer resolution failed! {}", err),
+      })
+    })?;
+    Ok(resolved_doc)
+  }
+}
+
+mod tests {
```
Comment (PhilippGackstatter) [inline identity_resolver/src/resolution/domain_linkage_verifier.rs:204]:
```suggestion
#[cfg(test)]
mod tests {
```

File: identity_credential/src/credential/did_configuration_resource.rs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+  fn builder() {
+    let issuer = Issuer::Url(Url::parse("did:example:issuer").unwrap());
+    let credential: Credential = DomainLinkageCredentialBuilder::new()
+      .issuance_date(Timestamp::now_utc())
+      .expiration_date(Timestamp::now_utc())
+      .issuer(issuer)
+      .origin(Url::parse("http://www.example.com").unwrap())
+      .build()
+      .unwrap();
+    let _config: DIDConfigurationResource = DIDConfigurationResource::new(vec![credential]);
```
Comment (PhilippGackstatter) [inline identity_credential/src/credential/did_configuration_resource.rs:216]: This line can be removed, I guess?

File: identity_credential/src/credential/did_configuration_resource.rs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+      .origin(Url::parse("http://www.example.com").unwrap())
+      .build();
+
+    assert!(matches!(credential, Err(Error::MissingExpirationDate)));
+  }
+
+  #[test]
+  fn from_json() {
+    const JSON1: &str = include_str!("../../tests/fixtures/well-known-configuration.json");
+    let _well_known: DIDConfigurationResource = DIDConfigurationResource::from_json(JSON1).unwrap();
```
Comment (PhilippGackstatter) [inline identity_credential/src/credential/did_configuration_resource.rs:257]: Nit: In these cases it's fine to:
```suggestion
    DIDConfigurationResource::from_json(JSON1).unwrap();
```

File: identity_credential/src/credential/did_configuration_resource.rs [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    let mut domains = OrderedSet::new();
+    domains.append(Url::parse("https://foo.example-1.com").unwrap());
+    domains.append(Url::parse("https://bar.example-2.com").unwrap());
+
+    let service: Service<CoreDID, Object> = DIDConfigurationResource::create_linked_domain_service(
+      CoreDIDUrl::parse("did:example:123#foo").unwrap(),
+      domains,
+    ).unwrap();
+
+    // Add a new Service.
```
Comment (PhilippGackstatter) [inline identity_credential/src/credential/did_configuration_resource.rs:271]:
```suggestion
```

File: identity_credential/src/credential/did_configuration_resource.rs [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    let service_from_json: Service<CoreDID, Object> = Service::from_json_value(json!({
+        "id":"did:example:123#foo",
+        "type": "LinkedDomains",
+        "serviceEndpoint": {
+          "origins": ["https://foo.example-1.com", "https://bar.example-2.com"]
```
Comment (PhilippGackstatter) [inline identity_credential/src/credential/did_configuration_resource.rs:276]: Nit: Would be slightly better to define these urls once and then include them here to prevent copy-paste mistakes.

File: examples/0_basic/8_domain_linkage.rs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  // The DID Configuration resource can be fetched using
+  // let configuration_resource: DIDConfigurationResource = DomainLinkageVerifier::fetch_configuration(Url::parse("https://foo.example.com/".to_owned())?).await?;
+  // But since we're working locally, we will simply deserialize the JSON.
+  let configuration_resource: DIDConfigurationResource =
+    DIDConfigurationResource::from_json(&configuration_resource_json)?;
+
+  // Verify the "fetched" DID Configuration resource.
+  let verification_result = verifier.verify_configuration(&configuration_resource).await;
+
+  // Verification returns ok
```
Comment (PhilippGackstatter) [inline examples/0_basic/8_domain_linkage.rs:113]:
```suggestion
  // Verification succeeds.
```

**Review by PhilippGackstatter — CHANGES_REQUESTED** [review]
Thanks for working on this! Really looking forward to this feature.

Currently, `DomainLinkageVerifier` lives in `identity_resolver`. However, if we can, I think that type should mirror the `PresentationValidator` and `CredentialValidator` types more. Those live in `identity_credential` and are injected with the necessary issuers and holder from the `Resolver` which is a nice separation of concerns. `identity_resolver` should resolve things, `identity_credential` should verify things, roughly speaking. (I'm going to slightly break that myself in the following suggestions).

I would suggest to

- move `DomainLinkageVerifier` to `identity_credential`. 
- change signature of `DomainLinkageVerifier::verify_domain_linkage_credential` to `DomainLinkageVerifier::verify_domain_linkage_credential<DOC: ValidatorDocument>(issuers: &[DOC], credential: &Credential, verifier_options: &VerifierOptions)` and make that method public. It does what `DomainLinkageVerifier::verify_domain_linkage_credential` currently does minus the resolution. The issuer documents are passed in, similar to `PresentationValidator` and `CredentialValidator`. That is handled either by the user themselves by calling `Resolver::resolve_issuers` or by a convenience method we can add: `Resolver::verify_domain_linkage_credential(credential: &Credential, verifier_options: &VerifierOptions)`.
  - I think it's fine to provide a method to verify all linked credentials in a configuration, but it should also be possible to verify an individual one as mentioned in the spec. Hence my suggestion to make `verify_domain_linkage_credential` public.
  - Since we're verifying credentials I think we should also take `VerifierOptions` here and pass that down.
- move `DomainLinkageVerifier::fetch_configuration` in `DIDConfigurationResource::fetch` but feature-gated so `identity_credential` does not depend on `reqwest`  by default.
- rename `DIDConfigurationResource` to `DomainLinkageConfiguration`. The spec name is pretty nondescript.

File: identity_resolver/src/resolution/domain_linkage_verifier.rs [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+  /// Verifies a [DID Configuration Resource](https://identity.foundation/.well-known/resources/did-configuration/#did-configuration-resource) for the following criteria:
+  /// - [Domain Linkage Credential](https://identity.foundation/.well-known/resources/did-configuration/#domain-linkage-credential)s
+  ///   are spec compliant and not expired.
+  /// - [DID Configuration Resource Verification](https://identity.foundation/.well-known/resources/did-configuration/#did-configuration-resource-verification)
+  ///
+  /// Note:
+  /// - Only [Linked Data Proof Format](https://identity.foundation/.well-known/resources/did-configuration/#linked-data-proof-format)
+  ///   is supported.
+  /// - The controller of the DID is not verified.
```
[the remark this answers is not in the document]
Comment (abdulmth) [reply identity_resolver/src/resolution/domain_linkage_verifier.rs:99]: In the [specs](https://identity.foundation/.well-known/resources/did-configuration/#did-configuration-resource-verification) there's a point about he controller: 

> a Verifier SHOULD consider the entity controlling the origin and the Controller of the DID to be the same entity.

This is not verified, maybe it should be more precise here.

[the remark this answers is not in the document]
Comment (PhilippGackstatter) [reply identity_resolver/src/resolution/domain_linkage_verifier.rs:99]: Thanks for the pointer. The spec just notes what assumptions we can make after successful verification, so I think we can remove that bullet point here.

File: identity_credential/src/credential/did_configuration_resource.rs [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+
+  /// List of extra properties (should be empty to be spec compliant).
+  pub fn properties_mut(&mut self) -> &mut Object {
+    &mut self.properties
+  }
+
+  /// Convenient function to create a spec compliant [Linked Domain Service Endpoint](https://identity.foundation/.well-known/resources/did-configuration/#linked-domain-service-endpoint)
+  pub fn create_linked_domain_service<D: DID>(
+    did_url: DIDUrl<D>,
+    domains: impl Into<OrderedSet<Url>>,
```
[the remark this answers is not in the document]
Comment (olivereanderson) [reply identity_credential/src/credential/did_configuration_resource.rs:86]: Although the Well Known DID Configuration spec allows for an array, the DID Core spec specifies that a [ service endpoint must be a string, a map or a set] hence I think it is correct to keep `OrderedSet<Url>` here.

**Review by olivereanderson — CHANGES_REQUESTED** [review]
Thanks for tackling this, it is indeed a very useful feature! 

In addition to @PhilippGackstatter's feedback I would suggest adding the following Error types to the `identity_credential` crate in the validator module .

```rust 
struct DomainLinkageCredentialVerificationError {
    // could also consider a vec of these if you want to aggregate all errors in a 
    // given credential, but it is probably fine to throw an error as soon as one is encountered. 
    cause: DomainLinkageCredentialVerificationErrorCause, 
    source: Option<Box<dyn std::error::Error + Send + Sync + 'static>>
}
// implement a "getter" for the cause field. 

#[non_exhaustive]
enum DomainLinkageCredentialErrorCause {
    MissingExpirationDate,
    ExceededExpirationDate,
    MissingOrigin,
    OriginMismatch, 
    ImpermissibleIdProperty, 
    IssuerSubjectMismatch,
    InvalidIssuanceDate,
    SignatureVerificationFailure,
}

struct DIDConfigurationResourceVerificationError {
   credential_idx_to_error: BtreeMap<usize, DomainLinkageCredentialVerificationError >
}
```
and these should be thrown  by the corresponding verification methods on the `DomainLinkageVerifier` .  The errors should implement the `Error` trait (at least those types whose name end in `Error`. For the `ErrorCause` it is really enough to just implement `Display`, but you might want to use `thiserror` for that in which case `Error` is also implemented). Note that the names I suggested are very long so please feel free to come up with better names!


In the `identity_resolver` crate you can extend the `ErrorCause` enum with the following  variants:
```
.. 
DomainLinkageVerification(DomainLinkageVerificationError), 
DIDResourceConfigurationVerification(DIDResourceConfigurationError),
ConfigurationResourceFetching(Box<dyn std::error::Error + Send + Sync + 'static>), 
}

```
and the `ResolutionAction` enum in that crate can also be extended with a variant giving a hint about the index of the domain linkage credential whose issuer could not be resolved similar to the existing `PresentationIssuersResolution(usize)` variant.

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+                return Err(DomainLinkageVerificationError {
+                  cause: DomainLinkageVerificationErrorCause::IssuerSubjectMismatch,
+                  source: None,
+                });
+              }
+            }
+            Err(_) => {
+              return Err(DomainLinkageVerificationError {
+                cause: DomainLinkageVerificationErrorCause::InvalidSubjectId,
+                source: None,
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/domain_linkage_validator.rs:199]:
```suggestion
            Err(err) => {
              return Err(DomainLinkageVerificationError {
                cause: DomainLinkageVerificationErrorCause::InvalidSubjectId,
                source: Some(Box::new(err)),
```

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+          .ok_or(DomainLinkageVerificationError {
+            cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
+            source: None,
+          })?;
+        let origin_url = Url::parse(origin)
+          .map_err(|_err| Url::parse("https://".to_owned() + origin))
+          .map_err(|_err| DomainLinkageVerificationError {
+            cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
+            source: None,
+          })?;
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/domain_linkage_validator.rs:222]: This maps the error of the first `Url::parse` to the result of the second `Url::parse` which is not what we want. It also always retries to parse the Url no matter what error is returned.
Suggestion:

```suggestion
     let origin_url: Url = match Url::parse(origin) {
          Ok(url) => url,
          Err(identity_core::Error::InvalidUrl(url::ParseError::RelativeUrlWithoutBase)) => {
            match Url::parse("https://".to_owned() + origin) {
              Ok(url) => url,
              Err(err) => {
                return Err(DomainLinkageVerificationError {
                  cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
                  source: Some(Box::new(err)),
                });
              }
            }
          }
          Err(err) => {
            return Err(DomainLinkageVerificationError {
              cause: DomainLinkageVerificationErrorCause::InvalidSubjectOrigin,
              source: Some(Box::new(err)),
            });
          }
        };
```

Requires `url = { version = "2.2", default-features = false }` in the `Cargo.toml`.

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+  use crate::credential::Credential;
+  use crate::credential::DomainLinkageConfiguration;
+  use crate::validator::domain_linkage_validator::DomainLinkageVerifier;
+  use crate::validator::CredentialValidationOptions;
+  use identity_core::common::Url;
+  use identity_core::convert::FromJson;
+  use identity_did::document::CoreDocument;
+
+  #[test]
+  pub(crate) fn test() {
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/domain_linkage_validator.rs:253]: Include more tests, use better test names :wink:.

File: identity_credential/src/validator/errors.rs [hunk 13]
```diff
@@ -139,4 +140,65 @@ impl Display for CompoundPresentationValidationError {
...
+}
+
+impl Display for DomainLinkageVerificationError {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    write!(f, "{:?}; ", self.cause)?;
+    match &self.source {
+      Some(err) => write!(f, "{}", err)?,
+      None => {}
+    }
+    Ok(())
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/errors.rs:167]:
```suggestion
    write!(f, "{}", self.cause)
```

Print the `Display` impl of the `cause`, not the `Debug` one.
Don't include the source error in the `Display` impl. Higher-level functionality/crates will do the formatting of these error chains. See also #1017.

File: identity_credential/src/validator/errors.rs [hunk 14]
```diff
@@ -139,4 +140,65 @@ impl Display for CompoundPresentationValidationError {
...
+  pub cause: DomainLinkageVerificationErrorCause,
+  /// Source of the error.
+  pub source: Option<Box<dyn Error + Send + Sync + 'static>>,
+}
+impl DomainLinkageVerificationError {
+  /// Provides the cause of the error.
+  pub fn get_cause(&self) -> &DomainLinkageVerificationErrorCause {
+    &self.cause
+  }
+}
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/errors.rs:158]: Is this really needed? The field is already public.

File: identity_credential/src/validator/errors.rs [hunk 15]
```diff
@@ -139,4 +140,65 @@ impl Display for CompoundPresentationValidationError {
   }
 }
 
-impl std::error::Error for CompoundPresentationValidationError {}
+impl Error for CompoundPresentationValidationError {}
+
+#[derive(Debug, thiserror::Error)]
+/// An error caused by a failure to verify a Domain Linkage configuration or credential.
+pub struct DomainLinkageVerificationError {
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/errors.rs:147]: Should we mark this `[non_exhaustive]`? Probably not, but just want to make sure. @olivereanderson any opinions?

File: identity_credential/tests/fixtures/did_configuration/issuer-did-document.json [hunk 16]
```diff
@@ -0,0 +1,11 @@
+{
+  "id": "did:iota:rms:0xe5d5a7c02874b0284ccc2f28bada68780cad26702e7e8186d521e9615a73bc45",
+  "verificationMethod": [
+    {
+      "id": "did:iota:rms:0xe5d5a7c02874b0284ccc2f28bada68780cad26702e7e8186d521e9615a73bc45#key-1",
+      "controller": "did:iota:rms:0xe5d5a7c02874b0284ccc2f28bada68780cad26702e7e8186d521e9615a73bc45",
+      "type": "Ed25519VerificationKey2018",
```
Comment (PhilippGackstatter) [inline identity_credential/tests/fixtures/did_configuration/issuer-did-document.json:7]: Is this file still needed when you use in-memory `CoreDocument`s for tests?

File: identity_credential/tests/fixtures/did_configuration/config1.json [hunk 17]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (PhilippGackstatter) [inline identity_credential/tests/fixtures/did_configuration/config1.json:1]: Consider giving the file a more descriptive name.

**Review by PhilippGackstatter — CHANGES_REQUESTED** [review]
Thanks for all the improvements, much appreciated, it's looking good overall!

A few comments on style, formatting, being good citizens in the Rust API world, errors, docs and other things.

I'm not too keen on including code in comments in the examples. They might go stale very fast, and then we have incorrect examples. My weak opinion is that we should rather describe what is commented.

File: identity_credential/src/credential/domain_linkage_configuration.rs [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /// List of the issuers of the Domain Linkage Credentials.
+  pub fn issuers(&self) -> Vec<Url> {
+    self
+      .linked_dids
+      .iter()
+      .map(|linked_did| linked_did.issuer.url().clone())
+      .collect()
+  }
```
Comment (olivereanderson) [inline identity_credential/src/credential/domain_linkage_configuration.rs:125]: I think Rust users would typically prefer an iterator to be returned here: 
```suggestion
  pub fn issuers(&self) -> impl Iterator<Item = &Url> {
    self
      .linked_dids
      .iter()
      .map(|linked_did| linked_did.issuer.url())
  }
```

File: identity_credential/src/validator/errors.rs (hunk 15 above)
Comment (olivereanderson) [reply identity_credential/src/validator/errors.rs:147]: Maybe it is best to instead make the fields private and introduce a getter for `cause` (the `source` is retrievable through the error trait).

File: identity_credential/src/validator/errors.rs [hunk 19]
```diff
@@ -139,4 +140,65 @@ impl Display for CompoundPresentationValidationError {
...
+    }
+    Ok(())
+  }
+}
+
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, strum::IntoStaticStr)]
+pub enum DomainLinkageVerificationErrorCause {
+  #[error("invalid credential")]
+  CredentialValidationError,
```
Comment (olivereanderson) [inline identity_credential/src/validator/errors.rs:175]: This should probably wrap `CompoundCredentialValidationError` or something like that to give more information about what went wrong when validating the attached credential.

File: identity_credential/src/credential/domain_linkage_configuration.rs [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+  /// Fetches the the DID Configuration resource via a GET request at the
+  /// well-known location: "`origin`/.well-known/did-configuration.json".
+  #[cfg(feature = "domain-linkage-fetch")]
+  pub async fn fetch_configuration(mut origin: Url) -> Result<DomainLinkageConfiguration> {
+    if origin.scheme() != "https" {
+      return Err(DomainLinkageError("origin` does not use `https` protocol".into()));
+    }
+    // todo: set file size limit to 1MB.
```
[the remark this answers is not in the document]
Comment (abdulmth) [reply identity_credential/src/credential/domain_linkage_configuration.rs:98]: I couldn't find a standard way to limit the size here with `reqwest` I'm not sure how to do this here, maybe use a stream if `reqwest` supports this?

File: identity_credential/src/credential/linked_domain_service.rs [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+
+    let service_from_json: Service<CoreDID, Object> = Service::from_json_value(json!({
+        "id":"did:example:123#foo",
+        "type": "LinkedDomains",
+        "serviceEndpoint": {
+          "origins": [domain_1, domain_2]
+        }
+    }))
+    .unwrap();
+    assert_eq!(service.into_service(), service_from_json);
```
[the remark this answers is not in the document]
Comment (PhilippGackstatter) [reply identity_credential/src/credential/linked_domain_service.rs]: If a conversion API is implemented, then implementing `From` and `Into` traits should always be preferred over an inherent function.

File: identity_credential/src/validator/errors.rs (hunk 13 above)
Comment (PhilippGackstatter) [reply identity_credential/src/validator/errors.rs]: You can try `cargo clean`, that sometimes helps with IDE problems.

File: identity_credential/src/credential/domain_linkage_configuration.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (PhilippGackstatter) [reply identity_credential/src/credential/domain_linkage_configuration.rs:98]: I think we can do it this way:

```rust
    // We use a stream so we can limit the size of the response to 1 MiB.
    let mut stream = client
      .get(origin.to_string())
      .send()
      .await
      .map_err(|err| DomainLinkageError(Box::new(err)))?
      .bytes_stream()
      .take(1_048_576);

    let mut json: Vec<u8> = Vec::new();
    while let Some(item) = stream.next().await {
      match item {
        Ok(bytes) => json.extend(bytes),
        Err(err) => return Err(DomainLinkageError(Box::new(err))),
      }
    }

    let domain_linkage_configuration: DomainLinkageConfiguration =
      DomainLinkageConfiguration::from_json_slice(&json).map_err(|err| DomainLinkageError(Box::new(err)))?;
```

File: identity_credential/src/validator/errors.rs (hunk 19 above)
Comment (olivereanderson) [reply identity_credential/src/validator/errors.rs:165]: From a logging perspective you are totally correct. 

The problem with the vague `CredentialValidation` variant is that one doesn't get an idea of which validations may fail by looking at the error. I am wondering whether it would make sense to either wrap `ValidationError` in the `CredentialValidation` variant (and call `CredentialValidator` with `FailFast::Yes` since domain linkage validation fails on the first error otherwise), or perhaps we include the relevant variants from  `ValidationError` as additional variants here. The latter is a bit more work to program, but would probably be nicer for TS users because more context becomes available with `Error.name`.  

Let me know what you prefer/think and also please free to chime in @PhilippGackstatter.

File: identity_credential/src/credential/domain_linkage_configuration.rs [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+/// Note:
+/// - Only [Linked Data Proof Format](https://identity.foundation/.well-known/resources/did-configuration/#linked-data-proof-format)
+///   is supported.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(try_from = "__DomainLinkageConfiguration")]
+pub struct DomainLinkageConfiguration(__DomainLinkageConfiguration);
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct __DomainLinkageConfiguration {
```
Comment (olivereanderson) [inline identity_credential/src/credential/domain_linkage_configuration.rs:38]: Does this type need to be public?

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 23]
```diff
@@ -0,0 +1,530 @@
...
+    credential: &Credential<T>,
+    domain: Url,
+    validation_options: &CredentialValidationOptions,
+  ) -> DomainLinkageValidationResult {
+    credential
+      .check_structure()
+      .map_err(|err| DomainLinkageValidationError {
+        cause: DomainLinkageValidationErrorCause::CredentialValidationError,
+        source: Some(Box::new(err)),
+      })?;
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/domain_linkage_validator.rs:87]: This is done in `CredentialValidator::validate` already.

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 24]
```diff
@@ -0,0 +1,530 @@
...
+      }
+    }
+  }
+
+  /// Validates a [Domain Linkage Credential](https://identity.foundation/.well-known/resources/did-configuration/#domain-linkage-credential).
+  ///
+  /// *`issuer`: issuer of the credential.
+  /// *`credential`: domain linkage Credential to be verified.
+  /// *`domain`: the domain hosting the credential.
+  pub fn validate_credential<T: Serialize, DOC: ValidatorDocument>(
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/domain_linkage_validator.rs:76]: Would be great if you could add some comments in this functions to each section of the validation. Because the function is so long it would be helpful to have a rough overview of what each section validates.
Alternatively, split the function into smaller functions with descriptive names.
(Or whatever you discuss with Oliver).

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 25]
```diff
@@ -0,0 +1,530 @@
...
+        let type_1 = types.get(0).ok_or(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
+          source: None,
+        })?;
+        let type_2: &String = types.get(1).ok_or(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
+          source: None,
+        })?;
+        let expected_types: HashSet<&str> = HashSet::from([
+          Credential::<T>::base_type(),
```
Comment (PhilippGackstatter) [inline identity_credential/src/validator/domain_linkage_validator.rs:126]: Checking whether this is in the credential is part of `CredentialValidator::check_structure` which is part of `CredentialValidator::validate` already.

**Review by PhilippGackstatter — APPROVED** [review]
Looks good to me, thanks a lot for your work on this!

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 26]
```diff
@@ -0,0 +1,530 @@
...
+          });
+        }
+      }
+      OneOrMany::Many(_) => {
+        return Err(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::MultipleCredentialSubjects,
+          source: None,
+        });
+      }
+    }
```
Comment (olivereanderson) [inline identity_credential/src/validator/domain_linkage_validator.rs:219]: I find it easier to read this if it is split up a bit more with some comments explaining the intention of each block. Here is a suggestion:

```suggestion
 // Extract credential subject
    let OneOrMany::One(ref credential_subject) = credential.credential_subject else {
      return Err(DomainLinkageValidationError {
        cause: DomainLinkageValidationErrorCause::MultipleCredentialSubjects,
        source: None,
      });
    };

    // Validate credential subject
    {
      let subject_id = credential_subject.id.as_deref().ok_or(DomainLinkageValidationError {
        cause: DomainLinkageValidationErrorCause::MissingSubjectId,
        source: None,
      })?;

      let subject_did = CoreDID::parse(subject_id.as_str()).map_err(|_| DomainLinkageValidationError {
        cause: DomainLinkageValidationErrorCause::InvalidSubjectId,
        source: None,
      })?;

      if issuer_did != subject_did {
        return Err(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::IssuerSubjectMismatch,
          source: None,
        });
      }
    }

    // Extract and validate origin
    {
      let origin: &str = credential_subject
        .properties
        .get("origin")
        .and_then(|value| value.as_str())
        .ok_or(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
          source: None,
        })?;

      let origin_url: Url = match Url::parse(origin) {
        Ok(url) => Ok(url),
        Err(identity_core::Error::InvalidUrl(url::ParseError::RelativeUrlWithoutBase)) => {
          Url::parse("https://".to_owned() + origin).map_err(|err| DomainLinkageValidationError {
            cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
            source: Some(Box::new(err)),
          })
        }
        Err(other_error) => Err(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
          source: Some(Box::new(other_error)),
        }),
      }?;

      if origin_url.origin() != domain.origin() {
        return Err(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::OriginMismatch,
          source: None,
        });
      }
    }
```

Comment (PhilippGackstatter) [reply identity_credential/src/validator/domain_linkage_validator.rs:219]: I think that's great, thanks for the comments, too!

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 27]
```diff
@@ -0,0 +1,485 @@
...
+          });
+        }
+      }
+      OneOrMany::One(_) => {
+        Err(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
+          source: None,
+        })?;
+      }
+    };
```
Comment (olivereanderson) [inline identity_credential/src/validator/domain_linkage_validator.rs:120]: Since `CredentialValidator::validate` already checks for the existence of the `VerifiableCredential` type, this can be simplified to the following:
```suggestion
   // Validate type.
    if !credential.types.iter().any(|type_| type_ == DomainLinkageConfiguration::domain_linkage_type()) {
      return Err(DomainLinkageValidationError {
        cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
        source: None,
      });
    }
```

File: identity_credential/src/validator/domain_linkage_validator.rs [hunk 28]
```diff
@@ -0,0 +1,485 @@
...
+  /// *`credential`: domain linkage Credential to be verified.
+  /// *`domain`: the domain hosting the credential.
+  pub fn validate_credential<T: Serialize, DOC: ValidatorDocument>(
+    issuer: &DOC,
+    credential: &Credential<T>,
+    domain: Url,
+    validation_options: &CredentialValidationOptions,
+  ) -> DomainLinkageValidationResult {
+    let issuer_did: CoreDID =
+      CoreDID::parse(credential.issuer.url().to_string()).map_err(|err| DomainLinkageValidationError {
```
Comment (olivereanderson) [inline identity_credential/src/validator/domain_linkage_validator.rs:82]: Nit: This will save an allocation (because `CoreDID` is not yet smart enough to re-use the owned string it is given): 
```
CoreDID::parse(credential.issuer.url().as_str()).map_err(|err| DomainLinkageValidationError {
```

**Review by olivereanderson — APPROVED** [review]
This looks great! 
Thank you for working on this cool feature :+1: 

I added two more comments that are mostly nits.
